File: apps/ios/ATTREQ/DesignSystem/Theme/Theme.swift
//
//  Theme.swift
//  ATTREQ
//
//  Semantic design tokens for the ATTREQ Redesign v2 handoff.
//  Source of truth: assets/design/ios-redesign-v2/attreq-shared.jsx
//  (ATTREQ_C light tokens / ATTREQ_DARK_C dark tokens).
//
//  Every color resolves through the asset catalog, so the system
//  appearance switch drives light/dark theming automatically.
//

import SwiftUI

// MARK: - Theme

/// Namespace mirroring the handoff token names (`bg`, `surface`, `t2`, `accent`, ...).
enum Theme {
    /// Screen background — light `#F5F2EE`, dark `#181512`.
    static let bg = Color("bg")
    /// Card / elevated surface — light `#FFFFFF`, dark `#231F1B`.
    static let surface = Color("surface")
    /// Deepest ink — light `#1C1917`, dark `#EDE9E3`.
    static let deep = Color("deep")
    /// Primary text — light `#1C1917`, dark `#EDE9E3`.
    static let text = Color("textPrimary")
    /// Secondary text — light `#78716C`, dark `#9A9088`.
    static let t2 = Color("textSecondary")
    /// Tertiary text — light `#A8A29E`, dark `#6E6862`.
    static let t3 = Color("textTertiary")
    /// Camel accent — light `#9B7B5A`, dark `#BA9272`.
    static let accent = Color("accent")
    /// Accent tint background — light 10%, dark 13% of accent.
    static let accentSoft = Color("accentSoft")
    /// Clay (destructive / skip) — light `#BF5C45`, dark `#D4705A`.
    static let clay = Color("clay")
    /// Clay tint background — light 10%, dark 12% of clay.
    static let claySoft = Color("claySoft")
    /// Moss (positive / worn) — light `#5A8A6A`, dark `#72AA86`.
    static let moss = Color("moss")
    /// Moss tint background — light 12%, dark 14% of moss.
    static let mossSoft = Color("mossSoft")
    /// Hairline border — 8% of ink.
    static let border = Color("border")
    /// Softer hairline border — 5% of ink.
    static let borderSoft = Color("borderSoft")
}

// MARK: - Garment placeholder gradients

/// The five garment-placeholder tones from the handoff's `garmentGrads`.
enum GarmentTone: CaseIterable {
    case top, bottom, outer, accent, shoes
}

extension Theme {
    /// Unit points approximating the CSS `linear-gradient(155deg, ...)` direction
    /// (CSS angles run clockwise from north; 155° points down and slightly right).
    private static let gradientStart = UnitPoint(x: 0.267, y: 0.0)
    private static let gradientEnd = UnitPoint(x: 0.733, y: 1.0)

    /// Two-stop placeholder gradient for a garment tone, matching
    /// `garmentGrads` in both appearances via the asset catalog.
    static func garmentGradient(_ tone: GarmentTone) -> LinearGradient {
        let name: String
        switch tone {
        case .top: name = "garmentTop"
        case .bottom: name = "garmentBottom"
        case .outer: name = "garmentOuter"
        case .accent: name = "garmentAccent"
        case .shoes: name = "garmentShoes"
        }
        return LinearGradient(
            colors: [Color("\(name)Start"), Color("\(name)End")],
            startPoint: gradientStart,
            endPoint: gradientEnd
        )
    }

    /// Card shadow from the handoff's `cardStyle` box-shadow:
    /// dark `0 2px 12px rgba(0,0,0,0.28)`, light `0 2px 8px rgba(0,0,0,0.04)`.
    /// CSS blur is twice the SwiftUI shadow radius, so 8px/12px blur maps to 4pt/6pt.
    static func cardShadow(isDark: Bool) -> (color: Color, radius: CGFloat, y: CGFloat) {
        isDark
            ? (color: Color.black.opacity(0.28), radius: 6, y: 2)
            : (color: Color.black.opacity(0.04), radius: 4, y: 2)
    }
}

// MARK: - Card background modifier

/// Applies the signature ATTREQ card treatment: surface background,
/// 20pt continuous corner radius, 1pt hairline border, and the
/// appearance-appropriate soft shadow.
struct AttreqCardBackground: ViewModifier {
    @Environment(\.colorScheme) private var colorScheme

    var padding: CGFloat = 16

    func body(content: Content) -> some View {
        let shape = RoundedRectangle(cornerRadius: 20, style: .continuous)
        let shadow = Theme.cardShadow(isDark: colorScheme == .dark)
        content
            .padding(padding)
            .background(Theme.surface, in: shape)
            .overlay(shape.strokeBorder(Theme.border, lineWidth: 1))
            .overlay {
                // Dark-mode inner top highlight from the handoff's `cardStyle`:
                // `inset 0 1px 0 rgba(255,255,255,0.04)` (same rim technique as
                // the tab bar's inner glow).
                if colorScheme == .dark {
                    RoundedRectangle(cornerRadius: 19, style: .continuous)
                        .strokeBorder(
                            LinearGradient(
                                colors: [Color.white.opacity(0.04), Color.white.opacity(0)],
                                startPoint: .top,
                                endPoint: .bottom
                            ),
                            lineWidth: 1
                        )
                        .padding(1)
                }
            }
            .shadow(color: shadow.color, radius: shadow.radius, x: 0, y: shadow.y)
    }
}

extension View {
    /// Wraps the view in the standard ATTREQ card style (see `AttreqCardBackground`).
    func attreqCard(padding: CGFloat = 16) -> some View {
        modifier(AttreqCardBackground(padding: padding))
    }
}

File: apps/mobile/src/design-system/theme/theme.ts
// ATTREQ design tokens — source of truth: assets/design/ios-redesign-v2/attreq-shared.jsx
// (ATTREQ_C light / ATTREQ_DARK_C dark), mirrored from apps/ios DesignSystem/Theme/Theme.swift.

export interface ThemeColors {
  bg: string;
  surface: string;
  deep: string;
  text: string;
  t2: string;
  t3: string;
  accent: string;
  accentSoft: string;
  border: string;
  borderSoft: string;
  clay: string;
  claySoft: string;
  moss: string;
  mossSoft: string;
}

export const lightColors: ThemeColors = {
  bg: '#F5F2EE',
  surface: '#FFFFFF',
  deep: '#1C1917',
  text: '#1C1917',
  t2: '#78716C',
  t3: '#A8A29E',
  accent: '#9B7B5A',
  accentSoft: 'rgba(155,123,90,0.10)',
  border: 'rgba(28,25,23,0.08)',
  borderSoft: 'rgba(28,25,23,0.05)',
  clay: '#BF5C45',
  claySoft: 'rgba(191,92,69,0.10)',
  moss: '#5A8A6A',
  mossSoft: 'rgba(90,138,106,0.12)',
};

export const darkColors: ThemeColors = {
  bg: '#181512',
  surface: '#231F1B',
  deep: '#EDE9E3',
  text: '#EDE9E3',
  t2: '#9A9088',
  t3: '#6E6862',
  accent: '#BA9272',
  accentSoft: 'rgba(186,146,114,0.13)',
  border: 'rgba(237,233,227,0.08)',
  borderSoft: 'rgba(237,233,227,0.05)',
  clay: '#D4705A',
  claySoft: 'rgba(212,112,90,0.12)',
  moss: '#72AA86',
  mossSoft: 'rgba(114,170,134,0.14)',
};

export type GarmentTone = 'top' | 'bottom' | 'outer' | 'accent' | 'shoes';

// Two-stop garment placeholder gradients (CSS linear-gradient(155deg, start, end)).
export const garmentGradients: Record<'light' | 'dark', Record<GarmentTone, [string, string]>> = {
  light: {
    top: ['#EDE7DF', '#DDD6CC'],
    bottom: ['#DAD4CC', '#CAC3BA'],
    outer: ['#E3DACE', '#D5CCBF'],
    accent: ['#F0EBE3', '#E5DED5'],
    shoes: ['#DFD9D2', '#D3CCC5'],
  },
  dark: {
    top: ['#3C3630', '#302A24'],
    bottom: ['#343030', '#28242A'],
    outer: ['#423A2C', '#362E22'],
    accent: ['#403C36', '#343028'],
    shoes: ['#3A3632', '#2E2A26'],
  },
};

export const spacing = {
  xs: 4,
  sm: 8,
  md: 12,
  lg: 16,
  xl: 20,
  '2xl': 24,
  '3xl': 32,
} as const;

export const radius = {
  xs: 8,
  sm: 12,
  md: 16,
  lg: 20,
  xl: 24,
  '2xl': 28,
  pill: 999,
} as const;

export interface CardShadow {
  shadowColor: string;
  shadowOpacity: number;
  shadowRadius: number;
  shadowOffset: { width: number; height: number };
  elevation: number;
}

// Handoff cardStyle box-shadow: light 0 2px 8px rgba(0,0,0,0.04), dark 0 2px 12px rgba(0,0,0,0.28).
export const cardShadow = (isDark: boolean): CardShadow =>
  isDark
    ? { shadowColor: '#000', shadowOpacity: 0.28, shadowRadius: 12, shadowOffset: { width: 0, height: 2 }, elevation: 5 }
    : { shadowColor: '#000', shadowOpacity: 0.04, shadowRadius: 8, shadowOffset: { width: 0, height: 2 }, elevation: 2 };

// Floating pill tab bar surface. Android has no cheap system-wide blur — semi-opaque approximation
// of the handoff's backdrop-blur(20) (documented divergence in docs/08-android-native/00-goal.md).
export const tabBarSurface = (isDark: boolean): string =>
  isDark ? 'rgba(24,21,18,0.96)' : 'rgba(245,242,238,0.95)';

export const tabActiveBg = (isDark: boolean): string =>
  isDark ? 'rgba(237,233,227,0.08)' : 'rgba(28,25,23,0.07)';

export interface Theme {
  colors: ThemeColors;
  isDark: boolean;
  garment: Record<GarmentTone, [string, string]>;
  spacing: typeof spacing;
  radius: typeof radius;
}

export const makeTheme = (isDark: boolean): Theme => ({
  colors: isDark ? darkColors : lightColors,
  isDark,
  garment: isDark ? garmentGradients.dark : garmentGradients.light,
  spacing,
  radius,
});
